src/mscorlib/shared/System/Threading/ExecutionContext.cs

```diff
@@ -42,6 +45,59 @@ internal void Undo(Thread currentThread)
                 ExecutionContext.Restore(currentThread, m_ec);
             }
         }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static bool CurrentIsDefault(Thread currentThread, ExecutionContext defaultContext)
+        {
+            // Current thread passed as parameter to avoid extern current native thread lookup call
+            Debug.Assert(currentThread == Thread.CurrentThread);
+            // Default context passed as parameter to avoid static inializer checks in NGen'd code
+            Debug.Assert(defaultContext == ExecutionContext.Default);
+
+            return (currentThread.ExecutionContext == defaultContext && currentThread.SynchronizationContext == null);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static void UndoToDefault(Thread currentThread, ExecutionContext defaultContext)
+        {
+            // Current thread passed as parameter to avoid extern current native thread lookup call
+            Debug.Assert(currentThread == Thread.CurrentThread);
+            // Default context passed as parameter to avoid static inializer checks in NGen'd code
+            Debug.Assert(defaultContext == ExecutionContext.Default);
+
+            // The common case is that these have not changed, so avoid the GC memory barrier 
+            // cost of an reference write if not needed.
+            if (currentThread.SynchronizationContext != null)
+            {
+                currentThread.SynchronizationContext = null;
+            }
+
+            if (currentThread.ExecutionContext != defaultContext)
+            {
+                // Restoring from a non-default context, need to check for AsyncLocal changes
+                ExecutionContext.RestoreDefault(currentThread, defaultContext);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static void SetDefaults()
+        {
+            // Current thread take to local to minimize extern current native thread lookup call
+            var currentThread = Thread.CurrentThread;
+            var defaultContext = ExecutionContext.Default;
+
+            // The common case is that these have not changed, so avoid the GC memory barrier 
+            // cost of an reference write if not needed.
+            if (currentThread.SynchronizationContext != null)
+            {
+                currentThread.SynchronizationContext = null;
+            }
+
+            if (currentThread.ExecutionContext != defaultContext)
+            {
+                currentThread.ExecutionContext = defaultContext;
```

Comment:
Currently if `Unsafe` changed async locals you'd wouldn't get notification at the end of the `Unsafe` call but you'd get them at the start of every regular threadpool item execution and the values would live in a kinda limbo land.

Also you could never move back onto the DefaultContext as even if you clear the values it won't move you back onto Default https://github.com/dotnet/coreclr/pull/8227#issue-190780413